File: web/src/perception.ts
import type { FieldObservation, PerceptionEvent } from "./schemas";
import { SCHEMA, SCHEMA_VERSION } from "./schemas";
import { uid } from "./hash";

export function makeSyntheticCsi(tick: number, body = "synthetic_cyd"): FieldObservation {
  const t = tick * 0.08;
  const rssi = -58 + 8 * Math.sin(t * 0.35);
  const csi: number[] = [];
  for (let i = 0; i < 32; i++) {
    const k = i / 31;
    let v =
      0.35 +
      0.25 * Math.sin(t + k * 6.28) +
      0.12 * Math.sin(t * 2.2 + k * 12) +
      0.05 * Math.sin(t * 7 + i);
    if (v < 0) v = 0;
    if (v > 1) v = 1;
    csi.push(v);
  }
  let sum = 0;
  let sq = 0;
  let peak = 0;
  for (const v of csi) {
    sum += v;
    sq += v * v;
    if (v > peak) peak = v;
  }
  const n = 32;
  const mean = sum / n;
  const energy = Math.sqrt(sq / n);
  let variance = 0;
  for (const v of csi) {
    const d = v - mean;
    variance += d * d;
  }
  const spread = Math.sqrt(variance / n);
  const rssi_n = Math.max(0, Math.min(1, (rssi + 90) / 60));
  return {
    body_id: body,
    body_type: "wifi_csi",
    timestamp: String(tick),
    synthetic: true,
    valid: true,
    rssi_dbm: rssi,
    csi,
    regions: [
      { name: "rssi", observed: rssi_n, confidence: 0.4 },
      { name: "csi_mean", observed: mean, confidence: 0.4 },
      { name: "csi_peak", observed: peak, confidence: 0.4 },
      { name: "csi_energy", observed: energy, confidence: 0.4 },
      { name: "csi_spread", observed: Math.min(1, spread * 2), confidence: 0.4 },
    ],
  };
}

export function observationToPerception(obs: FieldObservation, tick: number): PerceptionEvent {
  const energy = obs.regions.find((r) => r.name === "csi_energy")?.observed ?? 0;
  return {
    schema: SCHEMA.perception,
    version: SCHEMA_VERSION,
    id: uid("obs"),
    source: obs.body_id,
    timestamp: obs.timestamp,
    tick,
    modality: "csi",
    features: {
      rssi_dbm: obs.rssi_dbm,
      energy,
      mean: obs.regions.find((r) => r.name === "csi_mean")?.observed ?? 0,
      peak: obs.regions.find((r) => r.name === "csi_peak")?.observed ?? 0,
      spread: obs.regions.find((r) => r.name === "csi_spread")?.observed ?? 0,
    },
    confidence: 0.4,
  };
}

export function chatPerception(text: string, tick: number): PerceptionEvent {
  return {
    schema: SCHEMA.perception,
    version: SCHEMA_VERSION,
    id: uid("obs"),
    source: "throne-room",
    timestamp: new Date().toISOString(),
    tick,
    modality: "language",
    features: { text },
    confidence: 1,
  };
}

File: agent/perception.py
"""Perception adapters: synthetic CSI, FieldObservation → PerceptionEvent."""

from __future__ import annotations

import math
from datetime import datetime, timezone

from agent.hashutil import uid
from agent.schemas import FieldObservation, FieldRegion, PerceptionEvent


def _now() -> str:
    return datetime.now(timezone.utc).isoformat()


def make_synthetic_csi(tick: int, body: str = "synthetic_cyd") -> FieldObservation:
    t = tick * 0.08
    rssi = -58 + 8 * math.sin(t * 0.35)
    csi: list[float] = []
    for i in range(32):
        k = i / 31
        v = (
            0.35
            + 0.25 * math.sin(t + k * 6.28)
            + 0.12 * math.sin(t * 2.2 + k * 12)
            + 0.05 * math.sin(t * 7 + i)
        )
        csi.append(max(0.0, min(1.0, v)))
    n = 32
    mean = sum(csi) / n
    energy = math.sqrt(sum(v * v for v in csi) / n)
    variance = sum((v - mean) ** 2 for v in csi) / n
    spread = math.sqrt(variance)
    rssi_n = max(0.0, min(1.0, (rssi + 90) / 60))
    return FieldObservation(
        body_id=body,
        body_type="wifi_csi",
        timestamp=str(tick),
        synthetic=True,
        valid=True,
        rssi_dbm=rssi,
        csi=csi,
        regions=[
            FieldRegion("rssi", rssi_n, 0.4),
            FieldRegion("csi_mean", mean, 0.4),
            FieldRegion("csi_peak", max(csi), 0.4),
            FieldRegion("csi_energy", energy, 0.4),
            FieldRegion("csi_spread", min(1.0, spread * 2), 0.4),
        ],
    )


def observation_to_perception(obs: FieldObservation, tick: int) -> PerceptionEvent:
    energy = next((r.observed for r in obs.regions if r.name == "csi_energy"), 0.0)
    return PerceptionEvent(
        id=uid("obs"),
        source=obs.body_id,
        timestamp=obs.timestamp,
        tick=tick,
        modality="csi",
        features={
            "rssi_dbm": obs.rssi_dbm,
            "energy": energy,
            "mean": next((r.observed for r in obs.regions if r.name == "csi_mean"), 0.0),
            "peak": next((r.observed for r in obs.regions if r.name == "csi_peak"), 0.0),
            "spread": next((r.observed for r in obs.regions if r.name == "csi_spread"), 0.0),
            "synthetic": obs.synthetic,
        },
        confidence=0.4 if obs.synthetic else 0.9,
    )


def chat_perception(text: str, tick: int) -> PerceptionEvent:
    return PerceptionEvent(
        id=uid("obs"),
        source="throne-room",
        timestamp=_now(),
        tick=tick,
        modality="language",
        features={"text": text},
        confidence=1.0,
    )
